src/app/component/user/components/profile/calendar/calendar.component.ts

```diff
@@ -10,15 +15,120 @@ import { CalendarBaseComponent } from '@shared/components/calendar-base/calendar
 })
 export class CalendarComponent extends CalendarBaseComponent implements OnInit, OnDestroy {
 
-  constructor(public translate: TranslateService,
-              public languageService: LanguageService) {
-                super(translate, languageService);
-              }
+  public isHabitsPopUpOpen = false;
+  public selectedDay: number | string;
+  public habitsCalendarSelectedDate: string;
+  public calendarIcons = calendarIcons;
+  public isDayTracked: boolean;
+  public formatedDate: string;
+  public isHabitListEditable: boolean;
+  public isHabitChecked: boolean;
+  public isHabitEnrolled: boolean;
+  public currentDate: Date = new Date();
+  public habits: HabitAssignInterface[];
+  public daysCanEditHabits = 7;
+  public isFetching: boolean;
+  public allHabitsEnrolled: boolean;
+
+  @HostListener('document:click') clickout() {
+    if (this.isHabitsPopUpOpen) {
+      this.closePopUp();
+    }
+  }
+
+  constructor(
+    public translate: TranslateService,
+    public languageService: LanguageService,
+    public habitAssignService: HabitAssignService
+  ) {
+    super(translate, languageService, habitAssignService);
+  }
 
   ngOnInit() {
     this.bindDefaultTranslate();
     this.subscribeToLangChange();
     this.buildCalendar();
+    this.markCalendarDays(true, this.calendarDay);
   }
 
+  public checkIfFuture = (dayItem: CalendarInterface) => {
+    this.formatedDate = this.formatDate(true, dayItem);
+    if (this.currentDate.setHours(0, 0, 0, 0) >= new Date(this.formatedDate).setHours(0, 0, 0, 0)) {
+      this.toggleHabitsList(dayItem);
+    }
+  }
+
+  public toggleHabitsList(dayItem: CalendarInterface) {
+    this.isFetching = true;
+    this.isHabitsPopUpOpen = !this.isHabitsPopUpOpen;
+    this.checkHabitListEditable();
+    this.getActiveDateHabits(this.formatedDate);
+    this.selectedDay = dayItem.numberOfDate;
+    this.habitsCalendarSelectedDate = this.months[dayItem.month] + ' ' + dayItem.numberOfDate + ', ' + dayItem.year;
+    this.isDayTracked = !this.isDayTracked;
+  }
+
+  public checkHabitListEditable() {
+    this.isHabitListEditable = false;
+    if (this.currentDate.setHours(0, 0, 0, 0) - this.daysCanEditHabits * 24 * 60 * 60 * 1000 <=
+        new Date(this.formatedDate).setHours(0, 0, 0, 0)) {
+      this.isHabitListEditable = true;
+    }
+  }
+
+  public getActiveDateHabits(date: string) {
+    this.habitAssignService.getHabitAssignByDate(date).pipe(
+      take(1),
+      map((habits: HabitAssignInterface[]) => habits.sort((habit1, habit2) => (habit1.id > habit2.id) ? 1 : -1))
```

Comment:
Move this array "sort" operation to a separate statement.

Reply:
Done